File: discrete-math/practice-02/solve_pr2_tasks4_6.js
const fs = require("fs");
const path = require("path");

const outDir = __dirname;

const universe = [1, 2, 3, 4, 5, 6, 7, 8, 9];
const rows = {
  A: { items: [2, 5, 6, 7], cost: 3 },
  B: { items: [2, 8, 9], cost: 1 },
  C: { items: [1, 3, 7], cost: 2 },
  D: { items: [3, 5, 6, 9], cost: 2 },
  E: { items: [1, 3, 4, 8, 9], cost: 3 },
  F: { items: [1, 4, 6], cost: 1 },
  G: { items: [4, 5, 8], cost: 1 },
  H: { items: [3, 7], cost: 1 },
};

const rowNames = Object.keys(rows);

function arrayEq(a, b) {
  return a.length === b.length && a.every((x, i) => x === b[i]);
}

function union(names) {
  return [...new Set(names.flatMap((name) => rows[name].items))].sort((a, b) => a - b);
}

function rowCost(names) {
  return names.reduce((total, name) => total + rows[name].cost, 0);
}

function combinations(items, size, start = 0, prefix = [], result = []) {
  if (prefix.length === size) {
    result.push([...prefix]);
    return result;
  }
  for (let i = start; i <= items.length - (size - prefix.length); i += 1) {
    prefix.push(items[i]);
    combinations(items, size, i + 1, prefix, result);
    prefix.pop();
  }
  return result;
}

function isCover(names) {
  return arrayEq(union(names), universe);
}

function isIrredundant(names) {
  return isCover(names) && names.every((name) => !isCover(names.filter((item) => item !== name)));
}

function allCovers() {
  const covers = [];
  for (let size = 1; size <= rowNames.length; size += 1) {
    for (const combo of combinations(rowNames, size)) {
      if (isCover(combo)) covers.push(combo);
    }
  }
  return covers;
}

const covers = allCovers();
const irredundantCovers = covers.filter(isIrredundant);
const minCost = Math.min(...covers.map(rowCost));
const minCostCovers = covers
  .filter((cover) => rowCost(cover) === minCost)
  .sort((a, b) => a.length - b.length || a.join("").localeCompare(b.join("")));
const minLength = Math.min(...covers.map((cover) => cover.length));
const shortestCovers = covers
  .filter((cover) => cover.length === minLength)
  .sort((a, b) => rowCost(a) - rowCost(b) || a.join("").localeCompare(b.join("")));

const D = [1, 2, 3, 4, 5, 6];
const relation = [];
for (const m of D) {
  for (const n of D) {
    if ((3 * m + 1 - 2 * n) % 2 === 0) relation.push([m, n]);
  }
}
const matrix = D.map((m) => D.map((n) => (((3 * m + 1 - 2 * n) % 2 === 0) ? 1 : 0)));

function hasPair(a, b) {
  return relation.some(([x, y]) => x === a && y === b);
}

function classifyRelation() {
  const reflexive = D.every((x) => hasPair(x, x));
  const irreflexive = D.every((x) => !hasPair(x, x));
  const symmetric = relation.every(([a, b]) => hasPair(b, a));
  const antisymmetric = relation.every(([a, b]) => a === b || !hasPair(b, a));
  const asymmetric = irreflexive && relation.every(([a, b]) => !hasPair(b, a));
  const transitive = relation.every(([a, b]) =>
    relation
      .filter(([x]) => x === b)
      .every(([, c]) => hasPair(a, c)));
  return { reflexive, irreflexive, symmetric, antisymmetric, asymmetric, transitive };
}

const task6 = {
  variant59: {
    allPlacements: 3 ** 9,
    firstCarExactly3: combination(9, 3) * (2 ** 6),
    eachCarExactly3: factorial(9) / (factorial(3) ** 3),
    distribution4_3_2: factorial(9) / (factorial(4) * factorial(3) * factorial(2)) * factorial(3),
  },
  variant33: {
    lettersIntoBoxes: permutation(11, 5),
  },
};

function factorial(n) {
  let result = 1;
  for (let i = 2; i <= n; i += 1) result *= i;
  return result;
}

function combination(n, k) {
  return factorial(n) / (factorial(k) * factorial(n - k));
}

function permutation(n, k) {
  return factorial(n) / factorial(n - k);
}

const results = {
  task4: {
    variant: 9,
    D,
    predicate: "(3m + 1 - 2n) is even",
    simplifiedPredicate: "m is odd",
    relation,
    matrix,
    classification: classifyRelation(),
  },
  task5: {
    variant: 24,
    universe,
    rows,
    irredundantCovers: irredundantCovers.map((cover) => ({
      rows: cover,
      cost: rowCost(cover),
      length: cover.length,
    })),
    minCost,
    minCostCovers: minCostCovers.map((cover) => ({ rows: cover, cost: rowCost(cover), length: cover.length })),
    minLength,
    shortestCovers: shortestCovers.map((cover) => ({ rows: cover, cost: rowCost(cover), length: cover.length })),
  },
  task6,
};

const graphLines = ["graph LR"];
for (const x of D) graphLines.push(`  v${x}(("${x}"))`);
for (const [m, n] of relation) graphLines.push(`  v${m} --> v${n}`);
fs.writeFileSync(path.join(outDir, "task4_relation_graph.mmd"), graphLines.join("\n") + "\n", "utf8");

const text = [
  "Практична робота 2: контрольні результати",
  "",
  `Задача 4: |R| = ${relation.length}, предикат еквівалентний умові m непарне.`,
  `Класифікація: ${JSON.stringify(results.task4.classification)}`,
  "",
  `Задача 5: мінімальна ціна = ${minCost}, покриття: ${minCostCovers.map((c) => c.join("")).join(", ")}`,
  `Найкоротша довжина = ${minLength}, покриття: ${shortestCovers.map((c) => c.join("")).join(", ")}`,
  "",
  `Задача 6, варіант 59: ${JSON.stringify(task6.variant59)}`,
  `Задача 6, варіант 33: ${JSON.stringify(task6.variant33)}`,
  "",
].join("\n");

fs.writeFileSync(
  path.join(outDir, "pr2_tasks4_6_results.json"),
  JSON.stringify(results, null, 2),
  "utf8",
);
fs.writeFileSync(path.join(outDir, "pr2_tasks4_6_results.txt"), text, "utf8");

console.log(text);

File: discrete-math/practice-02/solve_pr2_tasks4_6.py
from __future__ import annotations

import itertools
import json
import math
import os
from pathlib import Path


OUT_DIR = Path(os.path.relpath(Path(__file__).parent, Path.cwd()))


UNIVERSE = list(range(1, 10))
ROWS = {
    "A": {"items": [2, 5, 6, 7], "cost": 3},
    "B": {"items": [2, 8, 9], "cost": 1},
    "C": {"items": [1, 3, 7], "cost": 2},
    "D": {"items": [3, 5, 6, 9], "cost": 2},
    "E": {"items": [1, 3, 4, 8, 9], "cost": 3},
    "F": {"items": [1, 4, 6], "cost": 1},
    "G": {"items": [4, 5, 8], "cost": 1},
    "H": {"items": [3, 7], "cost": 1},
}
ROW_NAMES = list(ROWS)
D = list(range(1, 7))


def union(names: list[str] | tuple[str, ...]) -> list[int]:
    values = set()
    for name in names:
        values.update(ROWS[name]["items"])
    return sorted(values)


def row_cost(names: list[str] | tuple[str, ...]) -> int:
    return sum(int(ROWS[name]["cost"]) for name in names)


def is_cover(names: list[str] | tuple[str, ...]) -> bool:
    return union(names) == UNIVERSE


def is_irredundant(names: tuple[str, ...]) -> bool:
    return is_cover(names) and all(
        not is_cover(tuple(item for item in names if item != name))
        for name in names
    )


def relation_pairs() -> list[list[int]]:
    return [
        [m, n]
        for m in D
        for n in D
        if (3 * m + 1 - 2 * n) % 2 == 0
    ]


def has_pair(relation: list[list[int]], a: int, b: int) -> bool:
    return any(x == a and y == b for x, y in relation)


def classify_relation(relation: list[list[int]]) -> dict[str, bool]:
    reflexive = all(has_pair(relation, x, x) for x in D)
    irreflexive = all(not has_pair(relation, x, x) for x in D)
    symmetric = all(has_pair(relation, b, a) for a, b in relation)
    antisymmetric = all(a == b or not has_pair(relation, b, a) for a, b in relation)
    asymmetric = irreflexive and all(not has_pair(relation, b, a) for a, b in relation)
    transitive = all(
        all(has_pair(relation, a, c) for x, c in relation if x == b)
        for a, b in relation
    )
    return {
        "reflexive": reflexive,
        "irreflexive": irreflexive,
        "symmetric": symmetric,
        "antisymmetric": antisymmetric,
        "asymmetric": asymmetric,
        "transitive": transitive,
    }


def main() -> None:
    covers: list[tuple[str, ...]] = []
    for size in range(1, len(ROW_NAMES) + 1):
        for combo in itertools.combinations(ROW_NAMES, size):
            if is_cover(combo):
                covers.append(combo)

    irredundant_covers = [cover for cover in covers if is_irredundant(cover)]
    min_cost = min(row_cost(cover) for cover in covers)
    min_cost_covers = sorted(
        [cover for cover in covers if row_cost(cover) == min_cost],
        key=lambda cover: (len(cover), "".join(cover)),
    )
    min_length = min(len(cover) for cover in covers)
    shortest_covers = sorted(
        [cover for cover in covers if len(cover) == min_length],
        key=lambda cover: (row_cost(cover), "".join(cover)),
    )

    relation = relation_pairs()
    matrix = [
        [1 if (3 * m + 1 - 2 * n) % 2 == 0 else 0 for n in D]
        for m in D
    ]
    classification = classify_relation(relation)

    task6 = {
        "variant59": {
            "allPlacements": 3**9,
            "firstCarExactly3": math.comb(9, 3) * (2**6),
            "eachCarExactly3": math.factorial(9) // (math.factorial(3) ** 3),
            "distribution4_3_2": math.factorial(9)
            // (math.factorial(4) * math.factorial(3) * math.factorial(2))
            * math.factorial(3),
        },
        "variant33": {
            "lettersIntoBoxes": math.factorial(11) // math.factorial(11 - 5),
        },
    }

    results = {
        "task4": {
            "variant": 9,
            "D": D,
            "predicate": "(3m + 1 - 2n) is even",
            "simplifiedPredicate": "m is odd",
            "relation": relation,
            "matrix": matrix,
            "classification": classification,
        },
        "task5": {
            "variant": 24,
            "universe": UNIVERSE,
            "rows": ROWS,
            "irredundantCovers": [
                {"rows": list(cover), "cost": row_cost(cover), "length": len(cover)}
                for cover in irredundant_covers
            ],
            "minCost": min_cost,
            "minCostCovers": [
                {"rows": list(cover), "cost": row_cost(cover), "length": len(cover)}
                for cover in min_cost_covers
            ],
            "minLength": min_length,
            "shortestCovers": [
                {"rows": list(cover), "cost": row_cost(cover), "length": len(cover)}
                for cover in shortest_covers
            ],
        },
        "task6": task6,
    }

    text = "\n".join(
        [
            "Практична робота 2: контрольні результати",
            "",
            f"Задача 4: |R| = {len(relation)}, предикат еквівалентний умові m непарне.",
            f"Класифікація: {json.dumps(classification, ensure_ascii=False, separators=(',', ':'))}",
            "",
            f"Задача 5: мінімальна ціна = {min_cost}, покриття: {', '.join(''.join(c) for c in min_cost_covers)}",
            f"Найкоротша довжина = {min_length}, покриття: {', '.join(''.join(c) for c in shortest_covers)}",
            "",
            f"Задача 6, варіант 59: {json.dumps(task6['variant59'], ensure_ascii=False, separators=(',', ':'))}",
            f"Задача 6, варіант 33: {json.dumps(task6['variant33'], ensure_ascii=False, separators=(',', ':'))}",
            "",
        ]
    )

    (OUT_DIR / "pr2_tasks4_6_results.json").write_text(
        json.dumps(results, ensure_ascii=False, indent=2),
        encoding="utf-8",
    )
    (OUT_DIR / "pr2_tasks4_6_results.txt").write_text(text, encoding="utf-8")
    print(text)


if __name__ == "__main__":
    main()
